src/aspire/source/image.py

```diff
@@ -1289,6 +1290,89 @@ def _populate_local_metadata(self):
         """
         return []
 
+    @staticmethod
+    def _prepare_relion_optics_blocks(metadata):
+        """
+        Split metadata into RELION>=3.1 style `data_optics` and `data_particles` blocks.
+
+        The optics block has one row per optics group with:
+        `_rlnOpticsGroup`, `_rlnOpticsGroupName`, and optics metadata columns.
+        The particle block keeps the remaining columns and includes a per-particle
+        `_rlnOpticsGroup` that references the optics block.
+        """
+        # Columns that belong in RELION's optics table.
+        all_optics_fields = [
+            "_rlnImagePixelSize",
+            "_rlnMicrographPixelSize",
+            "_rlnSphericalAberration",
+            "_rlnVoltage",
+            "_rlnAmplitudeContrast",
+            "_rlnImageSize",
+            "_rlnImageDimensionality",
+        ]
+
+        # We only write a data_optics block when every required field is present.
+        required_fields = [
+            "_rlnSphericalAberration",
+            "_rlnVoltage",
+            "_rlnAmplitudeContrast",
+            "_rlnImageSize",
+            "_rlnImageDimensionality",
+        ]
+        pixel_fields = ["_rlnImagePixelSize", "_rlnMicrographPixelSize"]
+
+        has_required = all(field in metadata for field in required_fields)
+        has_pixel_field = any(field in metadata for field in pixel_fields)
+
+        if not (has_required and has_pixel_field):
+            # Optics metadata incomplete, fall back to legacy single block.
+            logger.warning(
+                "Optics metadata incomplete, writing only data_particles block."
+            )
+            return None, metadata
+
+        # Restrict to the optics columns that are actually present on this source.
+        optics_value_fields = [
+            field for field in all_optics_fields if field in metadata
+        ]
+        n_rows = len(metadata["_rlnImageName"])
+
+        # Map each unique optics tuple to a 1-based group ID in order encountered.
+        group_lookup = OrderedDict()
+        optics_groups = np.empty(n_rows, dtype=int)
+
+        for idx in range(n_rows):
+            signature = tuple(metadata[field][idx] for field in optics_value_fields)
+            if signature not in group_lookup:
+                group_lookup[signature] = len(group_lookup) + 1
+            optics_groups[idx] = group_lookup[signature]
+
+        metadata["_rlnOpticsGroup"] = optics_groups
+
+        # Build the optics block rows and assign group names.
+        optics_block = OrderedDict()
+        optics_block["_rlnOpticsGroup"] = []
+        optics_block["_rlnOpticsGroupName"] = []
+        for field in optics_value_fields:
+            optics_block[field] = []
```

Comment:
This is good.  Try using a defaultdict(list) for `optics_block` and see if that reduces some of this code.

Fact check me, but I believe all Python dicts are now naturally ordered for versions of Python we support.